extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js

```diff
@@ -1,30 +1,45 @@
-import {Layout, applyFillContent} from '#core/dom/layout';
+import {Layout} from '#core/dom/layout';
 
-const TAG = 'amp-story-shopping-tag';
+import {Services} from '#service';
+
+import {StateProperty} from '../../amp-story/1.0/amp-story-store-service';
 
 export class AmpStoryShoppingTag extends AMP.BaseElement {
   /** @param {!AmpElement} element */
   constructor(element) {
     super(element);
 
-    /** @private {string} */
-    this.myText_ = TAG;
-
-    /** @private {?Element} */
-    this.container_ = null;
+    /** @private @const {!./amp-story-store-service.AmpStoryStoreService} */
+    this.storeService_ = null;
   }
 
   /** @override */
   buildCallback() {
     super.buildCallback();
-    this.container_ = this.element.ownerDocument.createElement('div');
-    this.container_.textContent = this.myText_;
-    this.element.appendChild(this.container_);
-    applyFillContent(this.container_, /* replacedContent */ true);
+    return Promise.all([
+      Services.storyStoreServiceForOrNull(this.win).then((storeService) => {
+        this.storeService_ = storeService;
+        this.storeService_.subscribe(
+          StateProperty.SHOPPING_STATE,
+          (shoppingState) => this.updateShoppingTag_(shoppingState)
+        );
+      }),
+    ]);
   }
 
   /** @override */
   isLayoutSupported(layout) {
     return layout === Layout.CONTAINER;
   }
+
+  /**
+   * @param {Object} shoppingState
```

Comment:
We need a typedef for this object.
That way we and future developers/reviewers can more easily understand the code.

Check out the example for interactives [in component](https://github.com/ampproject/amphtml/blob/main/extensions/amp-story-interactive/0.1/amp-story-interactive-abstract.js#L62-L89) in store service [here](https://github.com/ampproject/amphtml/blob/main/extensions/amp-story/1.0/amp-story-store-service.js#L53-L70) and [here](https://github.com/ampproject/amphtml/blob/main/extensions/amp-story/1.0/amp-story-store-service.js#L91-L92).

We can break that into a separate issue/PR if you want!

Reply:
Sounds like a plan - let's do it in another PR. Please file a separate ticket for the typedef and link it to this comment thread.

Reply:
Made an issue for this [here](https://github.com/ampproject/amphtml/issues/36911).